### ISSUE TITLE
BERT model for Machine Translation

### ISSUE THREAD

**Issue by KeremTurgutlu** [huggingface/pytorch-pretrained-BERT#31]
Is there a way to use any of the provided pre-trained models in the repository for machine translation task?

Thanks

**Comment by thomwolf** [comment]
Hi Kerem, I don't think so. Have a look at the fairsep repo maybe.

**Comment by JasonVann** [comment]
@thomwolf hi there, I couldn't find out anything about the fairsep repo. Could you post a link? Thanks!

**Comment by thomwolf** [comment]
Hi, I am talking about this repo: https://github.com/pytorch/fairseq.
Have a look at the section on using Transformer's for machine translation.

**Comment by alphadl** [comment]
I have conducted several MT experiments which fixed the embeddings by using BERT, **UNFORTUNATELY**, I find it makes performance worse. @JasonVann @thomwolf

**Comment by SinghJasdeep** [comment]
Hey! 

FAIR has demonstrated that using BERT for unsupervised translation greatly improves BLEU.

Paper: https://arxiv.org/abs/1901.07291

Repo: https://github.com/facebookresearch/XLM

Older papers showing pre-training with LM (not MLM) helps Seq2Seq: https://arxiv.org/abs/1611.02683

Hope this helps!

**Comment by gtesei** [comment]
These links are useful. 

Does anyone know if BERT improves things also for supervised translation?

Thanks.

**Comment by echan00** [comment]
> Does anyone know if BERT improves things also for supervised translation?

Also interested

**Comment by nyck33** [comment]
Because BERT is an encoder, I guess we need a decoder.  I looked here:  https://jalammar.github.io/
and it seems Openai Transformer is a decoder.  But I cannot find a repo for it.  
Does anyone know of a tutorial, blog or repo that shows how to build a decoder with attention mechanism?  
I thought about using the vector outputs of BERT as input to a sequence-to-sequence RNN model but would that possibly ignore some of the embedded information in the vector inputs?

**Comment by tacchinotacchi** [comment]
> These links are useful.
> 
> Does anyone know if BERT improves things also for supervised translation?
> 
> Thanks.

https://arxiv.org/pdf/1901.07291.pdf seems to suggest that it does improve the results for supervised translation as well. However this paper is not about using BERT embeddings, rather about pre-training the encoder and decoder on an Masked Language Modelling objective. The biggest benefit comes from initializing the encoder with the weights from BERT, and surprisingly using it to initialize the decoder also brings small benefits, even though if I understand correctly you still have to randomly initialize the weights for the encoder attention module, since it's not present in the pre-trained network.

**Comment by torshie** [comment]
I have managed to replace transformer's encoder with a pretrained bert encoder, however experiment results were very poor. It dropped BLEU score by about 4

The source code is available here: https://github.com/torshie/bert-nmt , implemented as a fairseq user model. It may not work out of box, some minor tweeks may be needed.

**Comment by rajarsheem** [comment]
I have been trying this: replacing the standard encoder in ONMT with BERT and other accompanying changes. Results are still not better than w/o BERT :/ 

[Here](https://github.com/OpenNMT/OpenNMT-py/compare/master...rajarsheem:bert_encoder) it is, in case anyone interested.

**Comment by sailordiary** [comment]
Could be relevant:

[Towards Making the Most of BERT in Neural Machine Translation](https://arxiv.org/pdf/1908.05672.pdf)
[On the use of BERT for Neural Machine Translation](https://arxiv.org/pdf/1909.12744.pdf)

**Comment by Bachstelze** [comment]
Also have a look at [MASS](https://github.com/microsoft/MASS) and [XLM](https://github.com/facebookresearch/XLM).

**Comment by lileicc** [comment]
Yes. It is possible to use both BERT as encoder and GPT as decoder and glue them together. 
There is a recent paper on this: Multilingual Translation via Grafting Pre-trained Language Models
https://aclanthology.org/2021.findings-emnlp.233.pdf
https://github.com/sunzewei2715/Graformer
